### DISCUSSION TITLE
Tasmota on end devices

### DISCUSSION THREAD

**Discussion by mikaelhertzman** [arendst/Tasmota#15113] [Q&A] [answered]
I have tried reading up on Tasmota, but having a hard to grasp the following concept. Is Tasmota for hubs only, or is it also used on end devices? 

For context I am trying to find a smart Zigbee light switch and use the options in Tasmota to customize/control its relay. Would this require me to flash Tasmota onto said smart switch, or would it be enough just to run Tasmota on the coordinator/hub?

**Comment by ascillato** [comment, marked as answer]
Tasmota is for Wi-fi devices. Not for ZigBee devices. Tasmota can connect and control ZigBee devices if where it is installed have a supported ZigBee chip (along with the ESP32 or ESP8266 WiFi chip) like the sonoff ZigBee bridge. Tasmota runs only in the ESP32 or ESP8266 wifi chips.
So, in the case of the sonoff ZigBee bridge, Tasmota can be used as a wifi to ZigBee bridge and it acts as a coordinator in the ZigBee network. And as a normal device in the wifi network.

So, with Tasmota in a ZigBee bridge you can control most of the ZigBee devices that supports the standard ZigBee protocol. There are some brands like tuya that is not using the standard ZigBee protocol, so those devices can be controlled only by tuya hubs.

In conclusion, Tasmota, in a WiFi to ZigBee bridge, can be used as a ZigBee coordinator to control your ZigBee devices with stock firmware (no modification needed on the zigbee end device) and control them through MQTT.

**Reply by mikaelhertzman** [reply]
Oh, I see. Thank you for a very detailed answer! (I did not know Wi-fi was required, I must have gotten confused by the tasmota-zigbee.bin firmware.) 

Does this mean that if I am running e.g. a Sonoff ZigBee Bridge with Tasmota on it, I will be able to use for example SetOption114 to decouple a smart switch, let's say the ZBMINI, that does _not_ have custom firmware (or Wi-fi either for that matter)? Or will the end device need to have support in its firmware for each custom option that Tasmota enables?

Let me try to make it more clear with an example - when I install the ZBMINI behind the light switch, I expect it not to do anything when I press it, except for sending a message via MQTT to the coordinator that will then be able to have logic in place for what should happen. By default all switches I have tried do indeed turn the relay on and off, so I want to disable that. Would Tasmota on a bridge allow me to modify the device behavior?

Sorry for long post, I really appreciate the help!

**Reply by sfromis** [reply]
No, when you have a Zigbee switch connected to a coordinator running Tasmota, the features for controlling local switches are not available, and especially not features which would be change the standard operation of the device in question, like SetOption114. You can make automations backend, or in the coordinator, but it all happens via Zigbee commands, not the regular Tasmota commands for non-Zigbee devices. For full detailed control of hardware, you should probably look to a non-Zigbee device (wifi), where Tasmota is able to directly control pins.

When you operate the physical switch input to a Zigbee switch, you should expect it to do the regular action, and then send a Zigbee message about what it did. In principle, it might be a feature to send Zigbee commands to make it decouple switch inputs from relay, but I'm not aware of this being a feature of such devices.

**Reply by mikaelhertzman** [reply]
> For full detailed control of hardware, you should probably look to a non-Zigbee device (WiFi), where Tasmota is able to directly control pins.

Have I understood you correct, that with the Sonoff Mini R2, Shelly or any other Wi-fi smart switch I would probably be able to control the relays? So the limitation is purely with Zigbee-only devices?

> When you operate the physical switch input to a Zigbee switch, you should expect it to do the regular action, and then send a Zigbee message about what it did.

Yes, this is what most smart switches do. However, Ubisys seem to sell switches that support "decoupling" the relay (at least according to the documentation on their website). The issue is that they are 100 USD a piece, while Sonoff or other options are more in the range of 20 USD.

I was really just trying to find a relatively cheap way of getting smart Zigbee switches and control them with Tasmota, but what I'm taking from this is that it's simply not possible without going for the WiFi options. What is it that makes all this possible with WiFi? Is it the higher bandwidth, or how the wireless protocol is built?

**Reply by sfromis** [reply]
Correct that with "classic" ESP8266/ESP32 based switches, where Tasmota gets flashed to the device, you should be able to treat control inputs and relay outputs independently, as Tasmota then is able to control all details of device operation. Strictly speaking, it is not "any other" as some of those devices are encumbered with a "TuyaMCU" which controls the device, with Tasmota merely being able to do the same as the app can do, no direct pin control. This is pretty similar to how it works with Zigbee devices.

Let me add that as long as you can get suitable Wifi-based hardware, and intend to control via Tasmota, there is no real benefit from "insisting" that communication is via Zigbee instead of Wifi.

What makes detailed control possible with Wifi-based switches is not Wifi, but the circumstance that Tasmota replaces the factory firmware on the device, allowing full control of every detail, possibly very different from factory firmware, including options for local automation. As Zigbee-based devices does not have a "Tasmota-compatible" MCU, you do not get the option of replacing factory firmware (at least not with Tasmota), and thus you are "stuck" with the factory firmware.

**Reply by mikaelhertzman** [reply]
> there is no real benefit from "insisting" that communication is via Zigbee instead of Wifi.

Yeah, running WiFi switches and devices would save me a lot of headache but I am avoiding going full WiFi for three reasons
a) I don't want to saturate my WiFi with "noise" from smart devices (I know Zigbee is 2.4 GHz as well, but that is manageable through other means)
b) I am running a big system of 100+ devices, and every device would need its own IP which would require me modifying my network
c) As I understand it Zigbee is a mesh network, meaning router devices will transmit messages along the mesh (and to my knowledge WiFi devices _don't_ do this)

But I am not sure anymore, perhaps going WiFi is the best option.

> What makes detailed control possible with Wifi-based switches is not Wifi, but the circumstance that Tasmota replaces the factory firmware on the device, allowing full control of every detail, possibly very different from factory firmware, including options for local automation. As Zigbee-based devices does not have a "Tasmota-compatible" MCU, you do not get the option of replacing factory firmware (at least not with Tasmota), and thus you are "stuck" with the factory firmware.

Okay, so assuming I use hardware with ESP chips, for full control Tasmota would need to be flashed to the end device then? Because the reply below is what confused me, but I guess it's referring to Zigbee only devices, correct?

"In conclusion, Tasmota, in a WiFi to ZigBee bridge, can be used as a ZigBee coordinator to control your ZigBee devices with **stock firmware** (no modification needed on the zigbee end device) and control them through MQTT."

**TLDR**: Let me see if I have finally understood this 😛,  for full functionality it's still necessary to flash Tasmota on the end device (assuming it has compatible hardware i.e. an ESP chip), correct?

**Reply by sfromis** [reply]
For extended functionality, compared to stock firmware from the manufacturer, you'll indeed need to replace firmware on the end device. Tasmota runs on ESP8266 and on several ESP32 variants, which is not used in Zigbee-based devices. In general, Zigbee-based devices have no option for Wifi too, except possible by getting a similar but different SKU from the manufacturer, replacing Zigbee support with Wifi.

The very, very low amount of Wifi traffic from Tasmota devices means that to me worrying about "noise" is simply a non-issue, unless you cram thousands of devices into your house. 100+ Tasmota devices is by far not a "big system" for wifi traffic generated. IP addresses is of course an issue to be dealt with, and crap ISP routers may not allow many, but with good full-house Wifi coverage (having a mesh) I do not feel a need to switch to Zigbee for the mesh you get with mains-powered Zigbee devices.

In my book, with Tasmota Wifi is preferable due to the option of full device control, and better integrated with Tasmota features. When I recommend Zigbee, it is mainly to people wanting cloudless local control, but wanting to avoid the sometimes tricky process of replacing factory firmware (or even wifi boards) in each device to be able to deploy Tasmota to the end device. 

For basic control, with the features the manufacturer put into the device, Zigbee is fine to use, but a bit less straightforward for automation. When running on a Zigbee coordinator bridge device, Tasmota is mainly bridging Zigbee traffic to special MQTT messages, which you can use for automation, including to some extent on the coordinator device itself (especially if it is an ESP32-based one with additional capabilities). If you want to do **all** your smarthome control in Home Assistant, Tasmota also has a ZHA mode allowing this component of HA to take full control of the Zigbee chip, instead of using MQTT.

**Reply by mikaelhertzman** [reply]
I agree, I am starting to realize that I may have to rethink this whole Zigbee-only plan I have been holding on to. Sometimes it's just difficult to let go of something once you have made a decision.

I am very grateful for all the great input you've given me, it has helped me a lot. Take care and stay safe!
